### PR TITLE
generator: Derive slice getters/setters for runtime-bounded static arrays

### PR DESCRIPTION
**Pull request by MarijnS95**
Upstream Vulkan is okay with the request to annotate static arrays with a `len="structField"` annotation when the size of the static arrary is bounded by a value at runtime.  This allows us to generate more convenient builder functions that copy slices into the target "builder" struct while _also_ updating the length, rather than forcing the caller to move an array of the desired length with zeroed elements and setting the length field separately.

In addition provide a safe getter (and use it in the `Debug` implementation) to return a slice view containing only valid items per the length field.  As with strings this is only possible for static-sized arrays as we can never safely dereference a random pointer.

### CODE REVIEW

File: ash/src/vk/definitions.rs [hunk 1]
```diff
@@ -1467,24 +1477,24 @@ impl ::std::default::Default for PhysicalDeviceMemoryProperties {
 }
 impl PhysicalDeviceMemoryProperties {
     #[inline]
-    pub fn memory_type_count(mut self, memory_type_count: u32) -> Self {
-        self.memory_type_count = memory_type_count;
+    pub fn memory_types(mut self, memory_types: &'_ [MemoryType]) -> Self {
```
Comment (MarijnS95) [inline ash/src/vk/definitions.rs:1480]: I've modified `vk.xml` on this struct as an example, but I believe we can merge this without any generator output changes while making sure the upstream CI doesn't start failing when they make and merge this change :ok_hand:

Comment (Ralith) [reply ash/src/vk/definitions.rs:1480]: SGTM.

**Review by Ralith — COMMENTED** [review]
Nice improvement!

File: ash/src/vk.rs
[the remark this answers is not in the document]
Comment (MarijnS95) [reply ash/src/vk.rs:1]: https://github.com/ash-rs/ash/pull/859

Comment (Ralith) [inline ash/src/vk.rs:1]: Still got a diff here.

File: ash/src/vk/definitions.rs (hunk 1 above)
Comment (Ralith) [reply ash/src/vk/definitions.rs:1480]: Approving conditional on this diff being rolled back.

**Review by Ralith — APPROVED** [review]

File: ash/src/vk/definitions.rs (hunk 1 above)
Comment (MarijnS95) [reply ash/src/vk/definitions.rs]: CI would reject those diffs anyway :tada:

File: generator/src/lib.rs [hunk 2]
```diff
@@ -2032,6 +2047,33 @@ fn derive_setters(
 
                 let mut slice_param_ty_tokens = field.safe_type_tokens(quote!('a), type_lifetime.clone(), None);
 
+                // vkxml wrongly annotates static arrays with a len= field is dynamic. These fields have a static length,
```
Comment (filnet) [inline generator/src/lib.rs:2050]: should be "as dynamic" instead of "is dynamic" ?

Comment (MarijnS95) [reply generator/src/lib.rs:2050]: Neither; I'll fix this up in another PR.

**Pull request merged**